src/main/java/io/vertx/core/streams/impl/PipeImpl.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2011-2017 The original author or authors
+ * ------------------------------------------------------
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Apache License v2.0 which accompanies this distribution.
+ *
+ *     The Eclipse Public License is available at
+ *     http://www.eclipse.org/legal/epl-v10.html
+ *
+ *     The Apache License v2.0 is available at
+ *     http://www.opensource.org/licenses/apache2.0.php
+ *
+ * You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.vertx.core.streams.impl;
+
+import java.util.Objects;
+import io.vertx.codegen.annotations.VertxGen;
+import io.vertx.core.VertxException;
+import io.vertx.core.streams.Pipe;
+import io.vertx.core.streams.Pump;
+import io.vertx.core.streams.ReadStream;
+import io.vertx.core.streams.WriteStream;
+
+/**
+ * 
+ * @author arnaud le roy
+ *
+ * @param <T>
+ */
+@VertxGen
+public class PipeImpl<T> implements Pipe<T> {
+
+	private ReadStream<T> _pipedStream;
+	private Pump pump;
+	private Boolean ended;
+
+	/**
+	 * 
+	 * @param rs
+	 */
+	public PipeImpl(ReadStream<T> rs) {
+		Objects.requireNonNull(rs);
+		this._pipedStream = rs;
+		this.ended = false;
+		this.pump = null;
+	}
+
+	/**
+	 * @param ws
+	 */
+	@Override
+	public Pipe<T> pipe(WriteStream<T> ws) {
```

Comment:
I think we should chain here, i.e we should have ability to pipe multiple times.

how do you handle the transform case with this API ?

Reply:
this is exemple of how i use it

    Pipe.create(filein).pipe(b64).pipe(filedest).start();

where 
1) upload is a AsyncFile, 
2) b64 implements readstream and writestream
3) filedest a AsyncFile.

this is the code of b64 class


```
package io.vertx.starter;
import java.util.Base64;
import java.util.Base64.Decoder;
import io.vertx.codegen.annotations.Nullable;
import io.vertx.core.Handler;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.streams.ReadStream;
import io.vertx.core.streams.WriteStream;

public class Base64Stream implements ReadStream<Buffer>, WriteStream<Buffer>{
	
	private Handler<Buffer> dataHandler;
	private Handler<Void> endHandler;
	private Handler<Throwable> exceptionHandler;
	private Buffer internalBuffer;
	private Decoder base64decoder;
	
	public Base64Stream() {
		internalBuffer = Buffer.buffer();
		base64decoder = Base64.getDecoder();
	}
	
	@Override
	public Base64Stream exceptionHandler(Handler<Throwable> handler) {
		this.exceptionHandler = handler;
		return this;
	}

	@Override
	public Base64Stream write(Buffer data) {
		internalBuffer.appendBuffer(data);
		int rest = internalBuffer.length() % 4;
		if (internalBuffer.length() >= 4) {
			byte[] _decoded = base64decoder.decode(internalBuffer.slice(0, internalBuffer.length()-rest).getBytes());
			if (dataHandler != null) {
				dataHandler.handle(Buffer.buffer(_decoded));
			}
		}
		
		internalBuffer = Buffer.buffer(internalBuffer.slice(internalBuffer.length()-rest,internalBuffer.length()).getBytes());
		return this;
	}

	@Override
	public void end() {
	    if (internalBuffer.length() > 0) {
	    	if (this.exceptionHandler != null) {
	    		this.exceptionHandler.handle(new Throwable("Bad base64 stream"));
	    	}
	    }
		if (endHandler != null) {
			endHandler.handle(null);
		}
	}

	@Override
	public Base64Stream setWriteQueueMaxSize(int maxSize) {
		return this;
	}

	@Override
	public boolean writeQueueFull() {
		return false;
	}

	@Override
	public Base64Stream drainHandler(@Nullable Handler<Void> handler) {
		return this;
	}

	@Override
	public Base64Stream handler(@Nullable Handler<Buffer> handler) {
		this.dataHandler = handler;
		return this;
	}

	@Override
	public Base64Stream pause() {
		return this;
	}

	@Override
	public Base64Stream resume() {
		return this;
	}

	@Override
	public Base64Stream endHandler(@Nullable Handler<Void> endHandler) {
		this.endHandler = endHandler;
		return this;
	}

}
```